server/sonar-process-monitor/src/main/java/org/sonar/application/SchedulerImpl.java

```diff
@@ -0,0 +1,278 @@
+    this.processWatcherDelayMs = l;
+    return this;
+  }
+
+  @Override
+  public void schedule() {
+    if (!nodeLifecycle.tryToMoveTo(NodeLifecycle.State.STARTING)) {
+      return;
+    }
+    processesById.clear();
+
+    for (ProcessId processId : AppSettingsHelper.getEnabledProcesses(settings)) {
+      SQProcess process = SQProcess.builder(processId)
+        .addProcessLifecycleListener(this)
+        .addEventListener(this)
+        .setWatcherDelayMs(processWatcherDelayMs)
+        .build();
+      processesById.put(process.getProcessId(), process);
+    }
+    operationalCountDown.set(processesById.size());
+    stopCountDown.set(processesById.size());
+
+    tryToStartAll();
+  }
+
+  private void tryToStartAll() {
+    try {
+      tryToStartEs();
+      tryToStartWeb();
+      tryToStartCe();
+    } catch (RuntimeException e) {
+      terminate();
+    }
+  }
+
+  private void tryToStartEs() {
+    tryToStartProcess(ProcessId.ELASTICSEARCH,
+      p -> true,
+      javaCommandFactory::createEsCommand);
+  }
+
+  private void tryToStartWeb() {
+    tryToStartProcess(ProcessId.WEB_SERVER,
+      p -> appState.isOperational(ProcessId.ELASTICSEARCH),
+      () -> javaCommandFactory.createWebCommand(true));
+  }
+
+  private void tryToStartCe() {
+    tryToStartProcess(ProcessId.COMPUTE_ENGINE,
+      p -> appState.isOperational(ProcessId.WEB_SERVER),
+      javaCommandFactory::createCeCommand);
+  }
+
+  private void tryToStartProcess(ProcessId processId, Predicate<SQProcess> startupCondition, Supplier<JavaCommand> commandSupplier) {
+    SQProcess process = processesById.get(processId);
+    if (process != null && startupCondition.test(process)) {
+      try {
+        process.start(() -> {
+          JavaCommand command = commandSupplier.get();
+          return javaProcessLauncher.launch(command);
+        });
+      } catch (RuntimeException e) {
+        // failed to start command -> stop everything
+        terminate();
+        throw e;
+      }
+    }
+  }
+
+  private void stopAll() {
+    // order is important for non-cluster mode
+    stopProcess(ProcessId.COMPUTE_ENGINE);
+    stopProcess(ProcessId.WEB_SERVER);
+    stopProcess(ProcessId.ELASTICSEARCH);
+  }
+
+  /**
+   * Request for graceful stop then blocks until process is stopped.
+   * Returns immediately if the process is disabled in configuration.
+   */
+  private void stopProcess(ProcessId processId) {
+    SQProcess process = processesById.get(processId);
+    if (process != null) {
+      process.stop(1, TimeUnit.MINUTES);
+    }
+
+  }
+
+  /**
+   * Blocks until all processes are stopped. Pending restart, if
+   * any, is disabled.
+   */
+  @Override
+  public void terminate() {
+    // disable ability to request for restart
+    restartRequested.set(false);
+    restartDisabled.set(true);
+
+    if (nodeLifecycle.tryToMoveTo(NodeLifecycle.State.STOPPING)) {
+      LOG.info("Stopping SonarQube");
+    }
+    stopAll();
+    // TODO cleanup ?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)